flow/environments/incite.py

```diff
@@ -207,4 +208,56 @@ def add_args(cls, parser):
         )
 
 
-__all__ = ["SummitEnvironment", "AndesEnvironment"]
+class CrusherEnvironment(DefaultSlurmEnvironment):
+    """Environment profile for the Cluster supercomputer.
+
+    https://docs.olcf.ornl.gov/systems/crusher_quick_start_guide.html
+    """
+
+    hostname_pattern = r".*\.crusher\.olcf\.ornl\.gov"
+    template = "crusher.sh"
+    cores_per_node = 128
+    gpus_per_node = 8
+    mpi_cmd = "srun"
+
+    @template_filter
+    def calc_num_nodes(cls, ngpus, ncpus, threshold):
+        """Compute the number of nodes needed to meet the resource request.
+
+        Also raise an error when the requested resource do not come close to saturating the asked
+        for nodes.
+        """
+        nodes_gpu = max(1, int(ceil(ngpus / cls.gpus_per_node)))
+        nodes_cpu = max(1, int(ceil(ngpus / cls.gpus_per_node)))
+        if nodes_gpu >= nodes_cpu:
+            check_utilization(nodes_gpu, ngpus, 8, threshold, "compute")
+            return nodes_gpu
+        check_utilization(nodes_cpu, ncpus, 64, threshold, "compute")
+        return nodes_cpu
+
+    @classmethod
+    def _get_mpi_prefix(cls, operation, parallel):
+        """Get the correct srun command for the job.
+
+        We don't currently support CPU/GPU mapping and expect the program to do this in code.
+        """
+        ngpus = operation.directives["ngpu"]
+        nranks = operation.directives.get("nranks", 1)
+        ncpus = max(
+            nranks * operation.directives.get("omp_num_threads", 1),
+            operation.directives.get("np", 1),
+        )
+        nodes = int(ceil(max(ngpus / cls.gpus_per_node, ncpus / cls.cores_per_node)))
+        base_str = f"{cls.mpi_cmd} -N{nodes} -n{nranks}"
+        if nranks == 1:
+            threads = max(
+                operation.directives.get("omp_num_threads", 1),
+                operation.directives.get("np", 1),
+            )
+        else:
+            threads = max(1, operation.directives.get("omp_num_threads", 1))
```

Comment:
And here you see one of the inconsistencies with `np` and `omp_num_threads`. If a user were to specify all three `np`, `nranks`, and `omp_num_threads`, they could get an unexpected combination of `-N`, `-n`, and `-c` options (e.g. `np=1024, nrank=10, omp_num_threads=64`) would request more nodes than necessary to run the given `srun` execution of 640 cores (10 ranks with 64 threads per rank).

Reply:
See comment above for some details on this question of how the variables relate. Specifying all three of `np`, `nranks`, `omp_num_threads` is probably incorrect in most user scripts, but we have not explicitly disallowed this.

More generally, I expect users to run with `--pretend` and ensure that the generated scripts make sense before blindly submitting. If the output doesn't match user expectations or if the scheduler rejects it, then users should file an issue or PR to fix it.